### PR TITLE
Tylermathis/ai/connectfour

### PR DESCRIPTION
**Pull request by TylerMathis**
Maybe I just suck at Connect Four, but I have yet to beat it :o

Check it out and try for yourself.
In order to recognize that you are playing a bot, the tagged user must have the @Chatbot role

### CODE REVIEW

File: src/app/plugins/connectfour.plugin.ts [hunk 1]
```diff
@@ -96,112 +100,214 @@ class ConnectFourGame {
...
 
   private _searchDX: number[] = [-1, -1, 0, 1, 1, 1, 0, -1];
   private _searchDY: number[] = [0, -1, -1, -1, 0, 1, 1, 1];
 
-  constructor(playerA: User, playerB: User) {
+  constructor(playerA: User, playerB: User, playingLion: Maybe<boolean>) {
     this._playerA = playerA;
     this._playerB = playerB;
 
+    this._playingLion = playingLion ? true : false;
```
Comment (c650) [inline src/app/plugins/connectfour.plugin.ts:119]: silly ternary.

File: src/app/plugins/connectfour.plugin.ts [hunk 2]
```diff
@@ -96,112 +100,214 @@ class ConnectFourGame {
...
+  private _getNormalizedDifficulty() {
+    return 1 / this._aiDifficulty;
+  }
+
+  // Evaluate the strength of the current board state,
+  // as it relates to Player 2's success.
+  private _evaluate(currentPlayer: number, depth: number) {
+    // If we have reached a win state, then the LAST move won.
+    if (this._checkWin()) {
+      return -4 * currentPlayer;
```
Comment (c650) [inline src/app/plugins/connectfour.plugin.ts:181]: -4?

File: src/app/plugins/connectfour.plugin.ts [hunk 3]
```diff
@@ -96,112 +100,214 @@ class ConnectFourGame {
...
+    }
+    return row;
+  }
+
+  private _longestChainOnBoard(currentPlayer?: number): number {
+    let longestChain = 0;
+    this._board.forEach((rowObj, rowIdx) => {
+      rowObj.forEach((_, colIdx) => {
+        longestChain = Math.max(
+          longestChain,
```
Comment (c650) [inline src/app/plugins/connectfour.plugin.ts:249]: use `reduce` so that you can do

```js
const longestchain = this._board.reduce(...)
```

**Review by c650 — APPROVED** [review]

File: src/app/plugins/connectfour.plugin.ts (hunk 3 above)
Comment (tanndlin) [reply src/app/plugins/connectfour.plugin.ts:249]: Better yet
```ts
return this._board.reduce(...)
```

File: src/app/plugins/connectfour.plugin.ts (hunk 1 above)
Comment (TylerMathis) [reply src/app/plugins/connectfour.plugin.ts:119]: Oops, looked like I dumped this change before I pushed.

File: src/app/plugins/connectfour.plugin.ts (hunk 2 above)
Comment (TylerMathis) [reply src/app/plugins/connectfour.plugin.ts:181]: Addressed with comment.

File: src/app/plugins/connectfour.plugin.ts (hunk 3 above)
Comment (TylerMathis) [reply src/app/plugins/connectfour.plugin.ts:249]: reduce will always return a value that is one dimension less than the object passed into it. How can I get around that for this case?

Comment (TylerMathis) [reply src/app/plugins/connectfour.plugin.ts:249]: As reference, this is what does not currently work.
`  private _longestChainOnBoard(currentPlayer?: number): number {
    return this._board.reduce((longestChain, rowObj, row) => {
      rowObj.forEach((_, col) => {
        longestChain = Math.max(
          longestChain,
          this._longestChainAtLocation(row, col, currentPlayer);
        );
      });
      return this._longestChainOnBoard;
    }, 0);`

Comment (tanndlin) [reply src/app/plugins/connectfour.plugin.ts:249]: > reduce will always return a value that is one dimension less than the object passed into it. How can I get around that for this case?

 Not necessarily, you can make it return whatever you want

Comment (tanndlin) [reply src/app/plugins/connectfour.plugin.ts:249]: > As reference, this is what does not currently work.
> 
> ```javascript
> 
> private _longestChainOnBoard(currentPlayer?: number): number {
> 
>     return this._board.reduce((longestChain, rowObj, row) => {
> 
>       rowObj.forEach((_, col) => {
> 
>         longestChain = Math.max(
> 
>           longestChain,
> 
>           this._longestChainAtLocation(row, col, currentPlayer);
> 
>         );
> 
>       });
> 
>       return this._longestChainOnBoard;
> 
>     }, 0);```

Foreach, doesn't return anything, you need a double reduce, or take the max of all 3 rows, so like
```ts 
return Math.max(this._board.map(row=> row.reduce((acc, rowObj, row) => Math.max(calculate the max, acc)
));
```

Comment (TylerMathis) [reply src/app/plugins/connectfour.plugin.ts]: Makes sense, reduce will always just feel useless to me I suppose ahahah

File: src/app/plugins/connectfour.plugin.ts [hunk 4]
```diff
@@ -96,112 +100,211 @@ class ConnectFourGame {
...
+    // We use a very simple evaluation: -(longest player one chain)
+    // This causes Lion to simply try to minimize Player 1's success,
+    // While our _checkWin case takes over when it's time to clinch victory.
+    if (depth === this._aiDepth) {
+      return this._longestChainOnBoard(1) - this._longestChainOnBoard(-1);
     }
-    // Subtract one to place piece one slot above obstruction.
+
+    const moves: number[] = [];
+    for (let col = 0; col < this._cols; col++) {
```
Comment (tanndlin) [inline src/app/plugins/connectfour.plugin.ts:197]: Can we use a foreach?

File: src/app/plugins/connectfour.plugin.ts [hunk 5]
```diff
@@ -96,112 +100,211 @@ class ConnectFourGame {
...
+    // Find first obstruction from top down in column.
+    let row = 0;
+    while (row < this._rows && this._board[row][col] === 0) {
+      row++;
+    }
+    return row;
+  }
+
+  private _longestChainOnBoard(currentPlayer?: number): number {
+    return this._board.reduce((longestChainOnBoard, rowObj, row) =>
```
Comment (tanndlin) [inline src/app/plugins/connectfour.plugin.ts:247]: This variable name is super long lmao, maybe `longestChain`

File: src/app/plugins/connectfour.plugin.ts [hunk 6]
```diff
@@ -96,112 +100,211 @@ class ConnectFourGame {
...
+          chains[direction]++;
         }
       }
+    }
 
-      // Return whether or not we found a chain of four.
-      return chains.includes(4);
-    };
+    // Return whether or not we found a chain of four.
+    return Math.max(...chains);
```
Comment (tanndlin) [inline src/app/plugins/connectfour.plugin.ts:271]: Do we need the spread operator? I thought `Math.max` could take in a `number[]` as is

Comment (TylerMathis) [reply src/app/plugins/connectfour.plugin.ts:271]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Math/max

File: src/app/plugins/connectfour.plugin.ts (hunk 4 above)
Comment (TylerMathis) [reply src/app/plugins/connectfour.plugin.ts:197]: Why? We only need an index here, and the index needs to be column oriented. In order to pull columns indices that are semantically correct out of a forEach we would need

```javascript
this._board.forEach((_row, row) => row.forEach((_col, col) => {
    // do the thing.
})
```

This takes more time, and more memory.


If I just misinterpreted what you meant let me know.

File: src/app/plugins/connectfour.plugin.ts (hunk 5 above)
Comment (TylerMathis) [reply src/app/plugins/connectfour.plugin.ts:247]: We have a `longestChainOnBoard` and a `longestChainAtLocation`.

For the sake of people extending this class I feel like having the distinction is important no?

File: src/app/plugins/connectfour.plugin.ts (hunk 4 above)
Comment (tanndlin) [reply src/app/plugins/connectfour.plugin.ts:197]: oh is it column oriented?
Then nevermind

File: src/app/plugins/connectfour.plugin.ts (hunk 5 above)
Comment (tanndlin) [reply src/app/plugins/connectfour.plugin.ts:247]: Fair enough

**Pull request merged**